src/middlewared/middlewared/plugins/init_shutdown_script.py

```diff
@@ -104,33 +107,85 @@ class Config:
         if verrors:
             raise verrors
 
+    async def execute_task(self, task):
+        task_type = task['type']
+        cmd = None
+
+        if task_type == 'COMMAND':
+            cmd = task['command']
+        elif os.path.exists(task['script'] or '') and os.access(task['script'], os.X_OK):
+            cmd = f'exec {task["script"]}'
+
+        if cmd:
+            proc = await Popen(
+                cmd,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                shell=True,
+                close_fds=True
+            )
+            await proc.wait()
+
+            if proc.returncode:
+                self.middleware.logger.debug(
+                    f'Execution failed for {task_type} {task["command"] if task_type == "COMMAND" else task["script"]}'
+                )
+
     @private
     @accepts(
         Str('when')
     )
-    def execute_init_tasks(self, when):
-        tasks = self.middleware.call_sync(
+    @job()
+    async def execute_init_tasks(self, job, when):
+
+        tasks = await self.middleware.call(
             'initshutdownscript.query', [
                 ['enabled', '=', True],
                 ['when', '=', when]
             ])
 
         for task in tasks:
-            task_type = task['type']
-            ret = None
-
-            if task_type == 'COMMAND':
-                ret = subprocess.run(
-                    task['command'],
-                    shell=True
-                )
-            elif os.path.exists(task['script']) and os.access(task['script'], os.X_OK):
-                ret = subprocess.run(
-                    f'exec {task["script"]}',
-                    shell=True
+            try:
+                await asyncio.wait_for(self.execute_task(task), timeout=task['timeout'])
+            except asyncio.TimeoutError:
+                self.middleware.logger.debug(
+                    f'{task["type"]} {task["command"] if task["type"] == "COMMAND" else task["script"]} timed out'
                 )
 
-            if ret and ret.returncode:
-                self.middleware.logger.debug(
-                    f'Execution failed for {task_type} {task["command"] if task_type == "COMMAND" else task["script"]}'
+        job.set_progress(100, f'Completed tasks for {when}')
+
+
+async def __event_system(middleware, event_type, args):
```

Comment:
Two things:

1) Reason I mentioned VMs is that we need a somewhat more generic shutdown handler, so we can later easily hook up other things (e.g. VMs, jails, whatever).

2) You get this event after same time shutdown/reboot is done (if delay = 0), so you have a race condition here in the shutdown process reading that sysctl right after rc.shutdown is executed and setting the sysctl here.
It is also possible shutdown was done manually via shell.

Reply:
Thank you @william-gr for pointing this out. Your concerns are absolutely valid. I had a thought about shutdown being done manually via shell. I don't think we have an efficient solution regarding that ( We should make sure that this is set via netcli though as we control that part ). But for the shell, if poweroff is invoked, it executes the rc.shutdown utility and any attempt to change the sysctl at that point is futile as the OS would still be holding on to the old value. Would you have any suggestions regarding that as to should we be handling this scenario ?

Reply:
You just need to be upfront about it. e.g. set the sysctl on system boot and call it again on events for init shutdown tasks changes.

Reply:
Thank you William. Yes that would indeed resolve this. Updating the PR